verifier/verifier.go

```diff
@@ -0,0 +1,142 @@
+/*
+ * ZCrypto Copyright 2017 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package verifier
+
+import (
+	"time"
+
+	"github.com/zmap/zcrypto/x509"
+)
+
+// VerificationResult contains the result of a verification of a certificate
+// against a browser. It discerns between NameError, meaning the name on the
+// certificate does not match, and a ValidationError, meaning some sort of issue
+// with the cryptography, or other issues with the chain.
+type VerificationResult struct {
+
+	// The name being checked, e.g. "www.censys.io".
+	Name string
+
+	// True if the certificate is whitelisted by a given browser as trusted, e.g.
+	// when Apple whitelisted subset of StartCom certs by SPKI hash in the wake of
+	// the WoSign incidents surrounding misissuance.
+	Whitelisted bool
+
+	// True is the certificate is blacklisted by a given browser as untrusted,
+	// e.g. Cloudflare certificates valid at the time of Heartbleed disclosure in
+	// Chrome.
+	Blacklisted bool
+
+	// ValiditionError will be non-nil when there was some sort of error during
+	// validation not involving a name mismatch, e.g. if a chain could not be
+	// built.
+	ValidationError error
+
+	// NameError will be non-nil when there was a mismatch between the name on the
+	// certificate and the name being verified against.
+	NameError error
+
+	// Chains is a list of validated certificate chains, starting at the
+	// certificate being verified, and ending at a certificate in the root store.
+	Chains [][]*x509.Certificate
```

Comment:
There may be multiple chains to the same root store through different intermediates, and you may want to apply different validation (e.g. remove any chain that contains SHA1).